src/util/AsioHelpers.h

```diff
@@ -1,65 +1,123 @@
 //   Copyright 2023, University of Freiburg,
 //   Chair of Algorithms and Data Structures.
-//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+//   Authors: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+//            Johannes Kalmbach <kalmbach@cs.uni-freiburg.de>
 
 #ifndef QLEVER_ASIOHELPERS_H
 #define QLEVER_ASIOHELPERS_H
 
+#include <absl/cleanup/cleanup.h>
+
+#include <boost/asio.hpp>
 #include <boost/asio/awaitable.hpp>
+#include <boost/asio/bind_cancellation_slot.hpp>
+#include <boost/asio/cancellation_signal.hpp>
+#include <boost/asio/co_spawn.hpp>
+#include <boost/asio/deferred.hpp>
 #include <boost/asio/post.hpp>
 #include <boost/asio/use_awaitable.hpp>
 
 #include "util/Exception.h"
+#include "util/Log.h"
 
 namespace ad_utility {
 
 namespace net = boost::asio;
 
-/// Helper function that ensures that co_await resumes on the executor
-/// this coroutine was co_spawned on.
-/// IMPORTANT: If the coroutine is cancelled, no guarantees are given. Make
-/// sure to keep that in mind when handling cancellation errors!
-template <typename T>
-inline net::awaitable<T> resumeOnOriginalExecutor(net::awaitable<T> awaitable) {
-  std::exception_ptr exceptionPtr;
-  try {
-    T result = co_await std::move(awaitable);
-    co_await net::post(net::use_awaitable);
-    co_return result;
-  } catch (...) {
-    exceptionPtr = std::current_exception();
-  }
-  auto cancellationState = co_await net::this_coro::cancellation_state;
-  if (cancellationState.cancelled() == net::cancellation_type::none) {
-    // use_awaitable always resumes the coroutine on the executor the coroutine
-    // was co_spawned on
-    co_await net::post(net::use_awaitable);
+// Internal implementation for `runFunctionOnExecutor`.
+namespace detail {
+// Call  `function()`. If this call doesn't throw, call `handler(nullptr,
+// resultOfTheCall)`, Else call `handler(current_exeption(), *nullptr)`. The
+// function is called directly, but the invocation of the handler is posted to
+// the associated executor of the `handler`, or to the `executor` if no such
+// associated executor exists.
+template <typename Executor, std::invocable Function, typename Handler>
+struct CallFunctionAndPassToHandler {
+  Executor executor_;
+  Function function_;
+  Handler handler_;
+  using Value = std::invoke_result_t<Function>;
+  static constexpr bool isVoid = std::is_void_v<Value>;
+  void operator()() {
+    // Get the executor of the completion handler, to which we will post the
+    // handler.
+    auto handlerExec = net::get_associated_executor(handler_, executor_);
```

Comment:
1. We don't have another fallback. For exception if the handler/token is a simple lambda. So in that case (which is probably rare) we just have to call the lambda. But maybe we cann statically assert that there is an `associated_executor` for the `handler_` to exclude this case.
2. The reposting of the handler to its executor is rather important, because otherwise, if the `executor_` is a strand and we use `use_awaitable` as the token which comes fron another strand, then the wrong strand is running after `co_awaiting` this whole thing (the unit test for the strands then fails). 
TLDR: I check if we can get rid of ANYH fallbacks here, but the posting is necessary.

Reply:
I would document this "fallback discrepancy" here in the code. Otherwise someone might stumble over this and declare it a bug or something in the future.